main-command/src/main/scala/sbt/internal/client/NetworkClient.scala

```diff
@@ -118,33 +235,104 @@ class NetworkClient(configuration: xsbti.AppConfiguration, arguments: List[Strin
+          attached.set(true)
+          Option(inputThread.get).foreach(_.drain())
+        }
+        pendingCompletions.remove(execId) match {
+          case null =>
+          case completions =>
+            completions(msg.result match {
+              case Some(o: JObject) =>
+                o.value
+                  .foldLeft(CompletionResponse(Vector.empty[String])) {
+                    case (resp, i) =>
+                      if (i.field == "items")
+                        resp.withItems(
+                          Converter
+                            .fromJson[Vector[String]](i.value)
+                            .getOrElse(Vector.empty[String])
+                        )
+                      else if (i.field == "cachedTestNames")
+                        resp.withCachedTestNames(
+                          Converter.fromJson[Boolean](i.value).getOrElse(true)
+                        )
+                      else if (i.field == "cachedMainClassNames")
+                        resp.withCachedMainClassNames(
+                          Converter.fromJson[Boolean](i.value).getOrElse(true)
+                        )
+                      else resp
+                  }
+              case _ => CompletionResponse(Vector.empty[String])
+            })
         }
-        ()
-      case _ =>
     }
   }
 
   def onNotification(msg: JsonRpcNotificationMessage): Unit = {
     def splitToMessage: Vector[(Level.Value, String)] =
       (msg.method, msg.params) match {
         case ("build/logMessage", Some(json)) =>
-          import sbt.internal.langserver.codec.JsonProtocol._
-          Converter.fromJson[LogMessageParams](json) match {
-            case Success(params) => splitLogMessage(params)
-            case Failure(e)      => Vector()
+          if (!attached.get) {
+            import sbt.internal.langserver.codec.JsonProtocol._
+            Converter.fromJson[LogMessageParams](json) match {
+              case Success(params) => splitLogMessage(params)
+              case Failure(_)      => Vector()
+            }
+          } else Vector()
+        case (`systemOut`, Some(json)) =>
+          Converter.fromJson[Seq[Byte]](json) match {
+            case Success(params) =>
+              if (params.nonEmpty) {
+                if (attached.get) {
+                  printStream.write(params.toArray)
+                  printStream.flush()
+                }
+              }
+            case Failure(_) =>
           }
+          Vector.empty
+        case (`promptChannel`, _) =>
+          batchMode.set(false)
+          Vector.empty
         case ("textDocument/publishDiagnostics", Some(json)) =>
           import sbt.internal.langserver.codec.JsonProtocol._
           Converter.fromJson[PublishDiagnosticsParams](json) match {
-            case Success(params) => splitDiagnostics(params)
-            case Failure(e)      => Vector()
+            case Success(params) => splitDiagnostics(params); Vector()
```

Comment:
Don't drop the diagnostics?

```suggestion
            case Success(params) => splitDiagnostics(params)
```

Reply:
Maybe? I have no idea why I made that change. It probably dates back to late December or early January.

Reply:
I made this change.